### ISSUE TITLE
PlaidHandler Load Time

### ISSUE THREAD

**Issue by shalbert94** [plaid/plaid-link-android#111]
# The problem

In the Plaid.com documentation for setting up an Android client, it says the following in the [Create a PlaidHandler](https://plaid.com/docs/link/android/#create-a-plaidhandler) section:

> Create a PlaidHandler - A PlaidHandler is a one-time use object used to open a Link session. It should be created as early as possible to warm-up Link so that it opens quickly. We recommend doing this as early as possible, since it must be completed before Link opens, and if delayed until just before Link opening it can have a perceptible impact on Link startup time.

After running the application with `delay`s before opening the Link UI, it's unclear what difference creating the PlaidHandler early has. Could you provide more details about how to incorporate this recommendation? 

# Environment

|                        |  |
| ------------------------| ---------- |
| Android OS Version                    | API 28   |
| Android Devices/Emulators    |        Pixel 2 emulator        |

# Steps to Reproduce

Open the [example/plaidLinkLoadTime](https://github.com/shalbert94/PlaidTest/tree/example/plaidLinkLoadTime) and follow the two "Todo"s

# Expected Result 
Increasing `delay` yields shorter load time for Link UI.

**Comment by amytang0** [comment]
Thanks for taking the time to setup the sample app; super helpful for understanding what you mean.

So what we envisioned when we separated the PlaidHandler initialization was that developers could call `PlaidHandler.create()` in the Activity creation, and then call `open` when user input occurs (i.e. when the button is pressed).

Under the hood, we are initializing several things in Plaid.create (e.g. a Retrofit object to do the server call). We found this to take a noticeable amount of time in our sample app, so separating the initialization from the server call to `openLink` ended up being a better user experience.

Since everyone's systems are different, I highly encourage you to move the `plaidHandler` creation to `MainActivity.onCreate` and see if you notice a difference between that and doing it in the `onClick` method.

Thank you!

**Comment by shalbert94** [comment]
Hi @amytang0, did that but not seeing a difference. Are you able to provide an example of this benefit?

**Comment by JeroenMols** [comment]
Hi @shalbert94 ,

Few thoughts that hopefully help you:

- We've already done tons of work to make the Plaid Link SDK open as fast as possible. And while initialization is indeed rather expensive like @amytang0  mentioned, it's nowhere near the order of seconds. Hence changing the delay from 1sec to 10sec won't yield different results.
- "Time to open" (or performance testing in general) is tricky to do well, as both the device hardware and the current CPU load of the device will affect the results
- If you aren't seeing any noticeable difference, that's good! 

> Note that if you're really interested in digging deeper into this subject, you could turn on [GPU profiling](https://developer.android.com/topic/performance/rendering/inspect-gpu-rendering#profile_rendering) and see how frame rendering is different for both cases. (initialize Plaid in onCreate() vs initialize Plaid in button click)

So in summary, we've made Plaid initialization as fast as possible so it can even be called from the `onClickHandler()`. However opening will be even faster if you already call initialize in `onCreate()`.

Hopefully this answers your question, I'm closing the issue in anticipation of that, feel free to reopen or file a new one if you have other issues/questions.

**Comment by shalbert94** [comment]
@JeroenMols, thank you. That explanation is very clear :)
